Implement UpdateDependencies script to automatically take new versions of dependencies

eerhardt opened the pull request:
There are 2 changes here:

1. Move dependency validation patterns to Item metadata and add programmatic names.
 - This allows automated tools to update the version numbers easier by referencing the names.
Another advantage is we can write regular expressions easier. Putting regular expressions in the Include attribute can cause a lot of issues.
2. Initial implementation of the UpdateDependencies.ps1 script. This script will be used to automate taking new versions of dependencies when new builds are available.
 - The first (and probably only) dependency to be automated is the CoreFX build numbers in the project.json files.

@dagood @weshaggard

eerhardt commented:
@weshaggard @dagood - I have updated the dir.props to follow the pattern we established in https://github.com/dotnet/buildtools/pull/557.  Also, I added the new update dependencies script.  PTAL.

/cc: @piotrpMSFT @stephentoub

weshaggard commented on UpdateDependencies.ps1:
```diff
@@ -0,0 +1,129 @@
+    $nsMgr = new-object Xml.XmlNamespaceManager $xml.NameTable
+    $nsMgr.AddNamespace('b', 'http://schemas.microsoft.com/developer/msbuild/2003')
+
+    $node = $xml.SelectSingleNode('//b:ValidationPattern[@Include = "CoreFxVersions"]/b:ExpectedPrerelease', $nsMgr)
```
Is this going to cause reformatting of dir.props? 

I thought we wanted to go ahead and put this into an independent property so it is a little easier to do a simple find replace. I suspect if we do that we don't even need use an XML API.

eerhardt replied:
The dir.props does get slightly reformatted by this, but there are only 2 changes. You can see what it would look like here:  https://github.com/eerhardt/corefx/commit/0693e38152333d6f0f6c8ba12b029f994a93a363#diff-0b192804a6349e8c26d2b027afbd89a2L53

There are only two minor changes:

```DIFF
-  <Import Project="$(MSBuildThisFileDirectory)Packaging.props"/>
+  <Import Project="$(MSBuildThisFileDirectory)Packaging.props" />
```

```DIFF
-    <DnuRestoreSource>@(DnuSourceList -> '--source %(Identity)', ' ')</DnuRestoreSource>
+    <DnuRestoreSource>@(DnuSourceList -&gt; '--source %(Identity)', ' ')</DnuRestoreSource>
```

My feeling was to not add a level of indirection (i.e. pull the value out into its own property) just for this script. Instead, write the dir.props as it makes sense, and then make the automation conform to it.

I can totally change this to to pull it out into its own variable, if you are strongly against this approach.

dagood replied:
I don't think I understood the comment on the other PR when I read it earlier, sorry--now I see that another reason to have it in its own property is that other parts of the build could use it later (without special code in the target). So it's not just for the update script's sake, it's for the build whenever we need this version number.

weshaggard replied:
Yes I would prefer we pull it out into its own property.

weshaggard commented on UpdateDependencies.ps1:
```diff
@@ -0,0 +1,129 @@
+    }
+}
+
+EnsureEnvironmentVariable "GITHUB_USER"
```
I would think it is better to pass these as parameters as opposed to using env variables. Is there a reason we need to use environment variables? Who's responsibility is to set those variable?

eerhardt replied:
The reason to use environment variables is because that is the easiest way to pass these variables in using a VSO build. The main runner of this script will be a VSO build definition, where all you need to do is define variables with the same names and these environment variables are set to the variables defined in VSO.

We do it this way for multiple CLI scripts that are run by VSO.

weshaggard replied:
OK I guess I will wait to see what the build definition actually looks like. For now I'm good with them being environment variables.

dagood commented on UpdateDependencies.ps1:
```diff
@@ -0,0 +1,129 @@
+
+    $UserName = $env:GITHUB_USER
+    $Email = $env:GITHUB_EMAIL
+    $CommitMessage = 'Updating dependencies from latest build numbers'
```
I think it might be useful to include the target version so it's easier to see in git history and shows up in the pull request title.

eerhardt commented:
@weshaggard and @dagood - I've responded to feedback and pushed the new changes. The dir.props replacement is a simple Regex now, so no formatting of the dir.props will occur. The new version number is written to the commit and PR messages.  PTAL.

weshaggard commented on UpdateDependencies.ps1:
```diff
@@ -0,0 +1,125 @@
+        return $false
+    }
+
+    $DirPropsPath = "$PSScriptRoot\dir.props"
```
I don't have a particular issue with what you have here but I thought I would at least share how to do the same thing in powershell cmdlets as well (I wrote this a while ago as when I was looking at trying to automate this).

```
$dirprops = Get-Content dir.props | % { $_ -replace "<CoreFxExpectedPrerelease>(.*)</CoreFxExpectedPrerelease>","<CoreFxExpectedPrerelease>$buildNumber</CoreFxExpectedPrerelease>" }
Set-Content dir.props $dirprops
```

eerhardt replied:
I changed this to use your code.

weshaggard commented on UpdateDependencies.ps1:
```diff
@@ -0,0 +1,125 @@
+# Updates all the project.json files with out of date version numbers
+function RunUpdatePackageDependencyVersions
+{
+    cmd /c $PSScriptRoot\build.cmd /t:UpdateInvalidPackageVersions | Out-Host
```
Any reason are you are calling cmd directly? You can just call the build.cmd command.

eerhardt replied:
I want to pass the fully qualified path to build.cmd, instead of relying on the 'working directory' being the repo root.  That's why I'm using $PSScriptRoot\build.cmd.

When I remove the `cmd /c` part, I get syntax errors with just

```ps1
$PSScriptRoot\build.cmd /t:UpdateInvalidPackageVersions | Out-Host
```

weshaggard replied:
Given you are constructing the string you might need to use `&` in front of the command, which tells PS to interpret the string as a command.

weshaggard commented on UpdateDependencies.ps1:
```diff
@@ -0,0 +1,125 @@
+# Creates a Pull Request for the updated version numbers
+function CreatePullRequest
+{
+    git add . | Out-Host
```
You could directly pass the `-a` option to your commit command and forgo the add.

eerhardt replied:
Fixed.

weshaggard commented on UpdateDependencies.ps1:
```diff
@@ -0,0 +1,125 @@
+# Updates the dir.props file with the latest CoreFX build number
+function UpdateValidDependencyVersionsFile
+{
+    if (!$CoreFxLatestVersion)
```
Any particular reason you are upper casing the first letter in the variable names?

eerhardt replied:
That is the convention used in all the CLI powershell scripts (which is where I learned to powershell):

https://github.com/dotnet/cli/blob/rel/1.0.0/scripts/run-build.ps1
https://github.com/dotnet/cli/blob/rel/1.0.0/scripts/obtain/install.ps1
https://github.com/dotnet/cli/blob/rel/1.0.0/scripts/common/_common.ps1
https://github.com/dotnet/cli/blob/rel/1.0.0/scripts/clean/clear-nuget-cache.ps1

weshaggard commented:
Just some minor nits/comments on the powershell script but otherwise LGTM. I will likely want to revisit the environment variable piece as some point.

dagood commented:
LGTM, and I see that we're already on a buildtools version with the validate changes this needs.

eerhardt commented on UpdateDependencies.ps1:
```diff
@@ -0,0 +1,125 @@
+
+    $env:GIT_COMMITTER_NAME = $UserName
+    $env:GIT_COMMITTER_EMAIL = $Email
+    git commit -m "$CommitMessage" --author "$UserName <$Email>" | Out-Host
```
We should check to see if anything was commited, and opt out if there were no changes.

dagood replied:
If you add `--porcelain` I think this is as simple as checking to see if there were any lines of output. (But maybe this comment is just to mark it as future work which is fine too.)

eerhardt replied:
Nice!  I used `git status --porcelain`.  Works like a charm.

eerhardt commented:
I've responded to feedback. PTAL

dagood commented:
Still LGTM. :+1:

weshaggard commented on UpdateDependencies.ps1:
```diff
@@ -0,0 +1,113 @@
+    [Parameter(Mandatory=$true)][string]$GitHubUser,
+    [Parameter(Mandatory=$true)][string]$GitHubEmail,
+    [Parameter(Mandatory=$true)][string]$GitHubPassword,
+    [string]$CoreFxVersionUrl='https://raw.githubusercontent.com/dotnet/versions/master/dotnet/corefx/master/Latest.txt', 
```
I assume this still needs to be actually defined.

eerhardt replied:
Yes.  This probably won't be the final location, given our conversation last week.  But the idea is that this will be an optional parameter going forward.  It felt wrong hard-coding 'https://github.com/eerhardt/versions' in the corefx repo, so I optimistically picked a URL.

Do you think it would be better to just make this mandatory now, since it doesn't work if you don't supply the actual value?  Or should I hard-code the path to 'eerhardt/versions' for now?  Or leave it as-is?

weshaggard replied:
Yes. I think this should be mandatory at least until we have a real value to default it to.

eerhardt replied:
Done.

weshaggard commented on dir.props:
```diff
@@ -68,20 +68,25 @@
   <PropertyGroup>
     <ValidatePackageVersions>true</ValidatePackageVersions>
     <ProhibitFloatingDependencies>true</ProhibitFloatingDependencies>
+    <CoreFxExpectedPrerelease>rc3-23921</CoreFxExpectedPrerelease>
```
Please update this to match the current value in dir.props before this gets merged.

eerhardt replied:
rebased.

weshaggard commented:
Fix the merges issues and update build values and then it looks good.

eerhardt commented:
I've rebased, squashed and pushed the latest.

I don't know why there aren't any CI jobs kicking off for this PR. If it was just the .ps1 script, I would be OK with merging it without CI. But since I'm updating the dir.props file, this should get a build to make sure it doesn't break the version validation script...

weshaggard commented:
Yes lets please verify CI works before merging. It looks like some jobs have triggered. 
\

eerhardt commented:
@dotnet-bot test this please

eerhardt commented:
@dotnet-bot test Innerloop OSX Debug Build and Test
@dotnet-bot test Innerloop Windows_NT Release Build and Test

eerhardt commented:
@dotnet-bot test Innerloop Windows_NT Release Build and Test

eerhardt commented:
@dotnet-bot test Innerloop Windows_NT Release Build and Test
